File: cli/internal/api/assets.go
package api

import (
	"bytes"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
	"net/textproto"
)

// UploadAsset PUTs an encrypted asset blob (whiteboard image binary) to
// /api/files/:fileID/assets/:assetID. The bytes must be already encrypted
// by crypto.EncryptAsset — server stores them opaque.
//
// Idempotent: re-PUTting the same (fileID, assetID) is a no-op for quota
// (server uses INSERT ON CONFLICT DO NOTHING). Caller can retry safely.
func (c *Client) UploadAsset(fileID, assetID string, ciphertext []byte) error {
	var buf bytes.Buffer
	w := multipart.NewWriter(&buf)

	h := make(textproto.MIMEHeader)
	h.Set("Content-Disposition", `form-data; name="file"; filename="asset"`)
	h.Set("Content-Type", "application/octet-stream")
	part, err := w.CreatePart(h)
	if err != nil {
		return err
	}
	if _, err := part.Write(ciphertext); err != nil {
		return err
	}
	w.Close()

	req, err := http.NewRequest(http.MethodPut,
		c.base+"/api/files/"+fileID+"/assets/"+assetID, &buf)
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", w.FormDataContentType())
	resp, err := c.do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("HTTP %d: %s", resp.StatusCode, string(body))
	}
	return nil
}

// DownloadAsset fetches the encrypted asset blob. Caller must
// crypto.DecryptAsset to recover the plaintext (the dataURL string for
// whiteboard images).
func (c *Client) DownloadAsset(fileID, assetID string) ([]byte, error) {
	resp, err := c.get("/files/" + fileID + "/assets/" + assetID)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("HTTP %d: %s", resp.StatusCode, string(body))
	}
	return io.ReadAll(resp.Body)
}

File: crates/kutup-cli/src/api/assets.rs
//! Whiteboard asset blobs — `PUT/GET /files/{id}/assets/{assetId}`.
//! Companion to `crates/kutup-server/src/handlers/file_assets.rs`.

use anyhow::Result;
use reqwest::blocking::multipart::{Form, Part};
use reqwest::Method;

use super::Client;

impl Client {
    /// Uploads an encrypted asset blob (multipart part name `file`).
    /// Idempotent server-side (`INSERT … ON CONFLICT DO NOTHING`).
    pub fn upload_asset(&self, file_id: &str, asset_id: &str, ciphertext: Vec<u8>) -> Result<()> {
        let part = Part::bytes(ciphertext)
            .file_name("asset")
            .mime_str("application/octet-stream")?;
        let form = Form::new().part("file", part);
        let resp = self
            .request(Method::PUT, &format!("/files/{file_id}/assets/{asset_id}"))
            .multipart(form)
            .send()?;
        super::check_ok(resp)
    }

    /// Downloads an encrypted asset blob.
    pub fn download_asset(&self, file_id: &str, asset_id: &str) -> Result<Vec<u8>> {
        let resp = self
            .request(Method::GET, &format!("/files/{file_id}/assets/{asset_id}"))
            .send()?;
        if resp.status().as_u16() >= 400 {
            return Err(super::api_error(resp));
        }
        Ok(resp.bytes()?.to_vec())
    }
}
